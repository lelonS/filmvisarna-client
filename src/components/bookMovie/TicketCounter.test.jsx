import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import TicketCounter from './TicketCounter';
import newDateFormat from '../../service/newDateFormat.js';

// Test data imports
import { movie, screening } from '../../../test-mocks/index.js';

// Mocks
const mockUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockUsedNavigate,
}));

let mockObject = { adult: 2, child: 0, senior: 0, total: 2 };
jest.mock('react-easier', () => ({
  ...jest.requireActual('react-easier'),
  useStates: () => mockObject
}));

window.scrollTo = jest.fn();

describe('TicketCounter', () => {

  beforeEach(() => {
    mockObject = {
      adult: 2,
      child: 0,
      senior: 0,
      total: 2
    };
  });

  test('renders correctly', () => {
    render(<TicketCounter seats={[]} movie={movie} screening={screening} />);
    expect(screen.getByText('Välj antal biljetter')).toBeInTheDocument();
    expect(screen.getByText(/Vuxen/i)).toBeInTheDocument();
    expect(screen.getByText(/Barn/i)).toBeInTheDocument();
    expect(screen.getByText(/Pensionär/i)).toBeInTheDocument();
    expect(screen.getByTestId('adult-counter')).toHaveTextContent('2');
    expect(screen.getByTestId('child-counter')).toHaveTextContent('0');
    expect(screen.getByTestId('senior-counter')).toHaveTextContent('0');
    expect(screen.getByAltText('movie poster')).toBeInTheDocument();
    expect(screen.getByAltText('movie poster').getAttribute('src')).toBe('/img/' + movie.img_poster);
    expect(screen.getByText(movie.title)).toBeInTheDocument();
    expect(screen.getByText(newDateFormat(screening.date) + ' | ' + screening.time)).toBeInTheDocument();
    expect(screen.getByText(screening.theaterName)).toBeInTheDocument();
    expect(screen.getByText(movie.genre)).toBeInTheDocument();
    expect(screen.getByText(movie.length)).toBeInTheDocument();
    expect(screen.getByText(movie.ageRestriction === 0 ? 'Ingen åldersgräns' : movie.ageRestriction + ' år')).toBeInTheDocument();
  });

  test('renders correctly with snapshot', () => {
    const tree = renderer.create(<TicketCounter seats={[]} movie={movie} screening={screening} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('increases counters correctly', () => {
    render(<TicketCounter seats={[]} setSeats={() => { }} />);

    const increaseButtons = screen.getAllByRole('button', { name: '+' });
    for (let i = 0; i < increaseButtons.length; i++) {
      fireEvent.click(increaseButtons[i]);
    }
    expect(mockObject.adult).toBe(3);
    expect(mockObject.child).toBe(1);
    expect(mockObject.senior).toBe(1);
    expect(mockObject.total).toBe(5);
  });

  test('decreases counters correctly', () => {
    render(<TicketCounter seats={[]} setSeats={() => { }} />);

    const decreaseButtons = screen.getAllByRole('button', { name: '-' });
    for (let i = 0; i < decreaseButtons.length; i++) {
      fireEvent.click(decreaseButtons[i]);
    }
    expect(mockObject.adult).toBe(1);
    expect(mockObject.child).toBe(0);
    expect(mockObject.senior).toBe(0);
    expect(mockObject.total).toBe(1);
  });

  test('counter does not increase when increase-button is clicked if total is 6', () => {
    mockObject = {
      adult: 3,
      child: 1,
      senior: 1,
      total: 5
    };
    render(<TicketCounter seats={[]} setSeats={() => { }} />);

    const increaseButtons = screen.getAllByRole('button', { name: '+' });
    for (let i = 0; i < increaseButtons.length; i++) {
      fireEvent.click(increaseButtons[i]);
    }
    expect(mockObject.adult).toBe(4);
    expect(mockObject.child).toBe(1);
    expect(mockObject.senior).toBe(1);
    expect(mockObject.total).toBe(6);
  });
});