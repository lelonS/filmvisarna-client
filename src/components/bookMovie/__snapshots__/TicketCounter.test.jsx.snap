// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketCounter renders correctly with snapshot 1`] = `
<div
  className="flex flex-col justify-evenly p-8 text-center text-xl text-white sm:mt-40 sm:flex-row sm:text-2xl lg:w-[80%]"
>
  <div
    className="sm:text-3xl lg:w-2/5"
  >
    <h2
      className="mb-6 text-3xl sm:text-3xl lg:text-left lg:text-[2.5rem]"
    >
      V채lj antal biljetter
    </h2>
    <div
      className="lg:flex lg:items-center lg:justify-between"
    >
      <p
        className="sm:text-2xl"
      >
        Vuxen (140 kr)
      </p>
      <div
        className=" flex items-center justify-center"
      >
        <button
          className="text-4xl sm:text-5xl"
          name="adult"
          onClick={[Function]}
        >
          -
        </button>
        <p
          className="mx-3 text-2xl sm:text-3xl"
          data-testid="adult-counter"
        >
          2
        </p>
        <button
          className="text-3xl sm:text-4xl"
          name="adult"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
    <div
      className="lg:flex lg:items-center lg:justify-between"
    >
      <p
        className="sm:text-2xl"
      >
        Barn (120 kr)
      </p>
      <div
        className="flex items-center justify-center"
      >
        <button
          className="text-4xl sm:text-5xl"
          name="child"
          onClick={[Function]}
        >
          -
        </button>
        <p
          className="mx-3 text-2xl sm:text-3xl"
          data-testid="child-counter"
        >
          0
        </p>
        <button
          className="text-3xl sm:text-4xl"
          name="child"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
    <div
      className="lg:flex lg:items-center lg:justify-between"
    >
      <p
        className="sm:text-2xl"
      >
        Pension채r (80 kr)
      </p>
      <div
        className="flex items-center justify-center"
      >
        <button
          className="text-4xl sm:text-5xl"
          name="senior"
          onClick={[Function]}
        >
          -
        </button>
        <p
          className="mx-3 text-2xl sm:text-3xl"
          data-testid="senior-counter"
        >
          0
        </p>
        <button
          className="text-3xl sm:text-4xl"
          name="senior"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
  </div>
  <div
    className="mt-6 flex gap-6"
  >
    <img
      alt="movie poster"
      className="h-44 sm:mt-auto sm:h-60"
      src="/img/bilar_poster.jpg"
    />
    <div
      className="flex flex-col justify-end text-start text-base"
    >
      <p
        className="text-xl"
      >
        Bilar 2
      </p>
      <p>
        Fredag 8:e december | 15:00
      </p>
      <p>
        Stora Salongen
      </p>
      <p>
        Familjefilm, Animation
      </p>
      <p>
        1 tim 45 min
         
      </p>
      <p>
         Ingen 책ldersgr채ns
      </p>
    </div>
  </div>
</div>
`;
